collect_app/src/main/java/org/odk/collect/android/activities/MainMenuActivity.java

```diff
@@ -640,4 +643,14 @@ private void disableSmsIfNeeded() {
                     .show();
         }
     }
+
+    private void setStorageMigrationBannerVisibility() {
+        if (!new StorageStateProvider().isScopedStorageUsed()) {
+            findViewById(R.id.storage_migration_banner).setVisibility(View.VISIBLE);
+        }
+    }
+
+    public void learnMoreAndMigrate(View view) {
+        new StorageMigratorTask(this).execute();
```

Comment:
I think it would be nice to have the execution encapsulated in a `ViewModel`. An example of this would be how `FormSaveViewModel#saveForm` works. Something like this for the `ViewModel`:

```java
class StorageMigrationViewModel {
    public LiveData<StorageMigratorResult> migrate() {
        MutableLiveData<StorageMigratorResult> result = new MutableLiveData<>();
        new StorageMigratorTask((migratorResult) -> result.setValue(migratorResult)).execute();
        return result;
    }
}
```

You'll see an example of a task with a listener like this in `FormSaveViewModel`.

In can then deal with the dialog/errors in the Activity:

```java
storageMigrationViewModel.migrate().observe(this, (result) -> {
    if (result != null) {
        // handle result
    }
});
```

Something like that lets us seperate out our UI logic from the migration logic and lets us avoid using `WeakReference` or worrying about `AsyncTask` memory leaks. Totally up for discussing if that approach doesn't feel right but it'd be nice for us to find a pattern we like for this kind of thing.

Reply:
I think it's a good idea I implemented it, thanks.